connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```diff
@@ -0,0 +1,349 @@
+            kafkaConsumer.close();
+        }
+    }
+
+    private void commitPendingAndWaitForTopics() throws InterruptedException {
+        //check if the last remaining topic was removed
+        if (kafkaConsumer.assignment().size() > 0) {
+            Map<TopicPartition, OffsetAndMetadata>
+                            offsetData =
+                            this.acknowlegmentHandler.removePartitions(kafkaConsumer.assignment());
+            commitOffsets(offsetData);
+            kafkaConsumer.unsubscribe();
+        }
+        synchronized (this) {
+            //recheck wait condition inside synchronized block
+            while (topics.size() == 0) {
+                //pause the polling thread until a topic comes in
+                log.debug("controlLoop waiting for topic");
+                this.wait();
+            }
+        }
+    }
+
+    private void commitPendingOffsetsForRemovedTopics() throws InterruptedException {
+        Set<TopicPartition> removed = new HashSet<>();
+        //synchronized to avoid letting the topic set change
+        synchronized (this) {
+            for (TopicPartition partition : kafkaConsumer.assignment()) {
+                if (!topics.contains(partition.topic())) {
+                    removed.add(partition);
+                }
+            }
+            topicsChanged = false;
+        }
+        //commit any removed partitions before we unsubscribe them
+        Map<TopicPartition, OffsetAndMetadata> offsetData = this.acknowlegmentHandler.removePartitions(removed);
+        commitOffsets(offsetData);
+    }
+
+    private ConsumerRecords<?, ?> flowControlledPoll() throws InterruptedException {
+        //pause partitions that haven't fully been processed yet and unpause those that have
+        Set<TopicPartition> paused = new HashSet<>();
+        Set<TopicPartition> unpaused = new HashSet<>();
+        for (Map.Entry<TopicPartition, AtomicInteger> entry : flowControl.entrySet()) {
+            if (entry.getValue().get() > 0) {
+                paused.add(entry.getKey());
+            } else {
+                unpaused.add(entry.getKey());
+            }
+        }
+        kafkaConsumer.pause(paused);
+        kafkaConsumer.resume(unpaused);
+        if (unpaused.size() == 0 && paused.size() > 0) {
+            synchronized (this) {
+                //wait for flowControl to notify us, resume after a short pause to allow for heartbeats
+                this.wait(100);
+            }
+            return null;
+        } else {
+            return kafkaConsumer.poll(100);
+        }
+    }
+
+    private void updateAssignment() {
+        this.acknowlegmentHandler.addPartitions(kafkaConsumer.assignment());
+        for (TopicPartition partition : kafkaConsumer.assignment()) {
+            this.flowControl.merge(partition, new AtomicInteger(),
+                                   (oldValue, newValue) -> oldValue == null ? newValue : oldValue);
```

Comment:
`oldValue` should never be null; otherwise the merge function wouldn't be called.

So, just use a `putIfAbsent`?